src/sfizz/FlexEnvelope.cpp

```diff
@@ -0,0 +1,206 @@
+   Note(jpc): implementation status
+
+- [ ] egN_points (purpose unknown)
+- [x] egN_timeX
+- [x] egN_levelX
+- [x] egN_shapeX
+- [x] egN_sustain
+- [ ] egN_dynamic
+- [ ] egN_loop
+- [ ] egN_loop_shape
+- [ ] egN_loop_count
+*/
+
+#include "FlexEnvelope.h"
+#include "FlexEGDescription.h"
+#include "Curve.h"
+#include "Config.h"
+#include "SIMDHelpers.h"
+#include <absl/types/optional.h>
+
+namespace sfz {
+
+struct FlexEnvelope::Impl {
+    const FlexEGDescription* desc_ { nullptr };
+    float samplePeriod_ { 1.0 / config::defaultSampleRate };
+    size_t delayFramesLeft_ { 0 };
+
+    //
+    float stageSourceLevel_ { 0.0 };
+    float stageTargetLevel_ { 0.0 };
+    float stageTime_ { 0.0 };
+    bool stageSustained_ { false };
+    const Curve* stageCurve_ { nullptr };
+
+    //
+    unsigned currentStageNumber_ { 0 };
+    float currentLevel_ { 0.0 };
+    float currentTime_ { 0.0 };
+    absl::optional<size_t> currentFramesUntilRelease_ { absl::nullopt };
+    bool isReleased_ { false };
+
+    //
+    void process(absl::Span<float> out);
+    bool advanceToNextStage();
+};
+
+FlexEnvelope::FlexEnvelope()
+    : impl_(new Impl)
+{
+}
+
+FlexEnvelope::~FlexEnvelope()
+{
+}
+
+void FlexEnvelope::setSampleRate(double sampleRate)
+{
+    Impl& impl = *impl_;
+    impl.samplePeriod_ = 1.0 / sampleRate;
+}
+
+void FlexEnvelope::configure(const FlexEGDescription* desc)
+{
+    Impl& impl = *impl_;
+    impl.desc_ = desc;
+}
+
+void FlexEnvelope::start(unsigned triggerDelay)
+{
+    Impl& impl = *impl_;
+    const FlexEGDescription& desc = *impl.desc_;
+
+    impl.delayFramesLeft_ = triggerDelay;
+
+    FlexEGPoint point;
+    if (!desc.points.empty())
+        point = desc.points[0];
+
+    //
+    impl.stageSourceLevel_ = 0.0;
+    impl.stageTargetLevel_ = point.level;
+    impl.stageTime_ = point.time;
+    impl.stageSustained_ = desc.sustain == 0;
+    impl.stageCurve_ = &point.curve();
+    impl.currentFramesUntilRelease_ = absl::nullopt;
+    impl.isReleased_ = false;
+
+    //
+    impl.currentStageNumber_ = 0;
+    impl.currentLevel_ = 0.0;
+    impl.currentTime_ = 0.0;
+}
+
+void FlexEnvelope::release(unsigned releaseDelay)
+{
+    Impl& impl = *impl_;
+    impl.currentFramesUntilRelease_ = releaseDelay;
+}
+
+void FlexEnvelope::process(absl::Span<float> out)
+{
+    Impl& impl = *impl_;
+    impl.process(out);
+}
+
+void FlexEnvelope::Impl::process(absl::Span<float> out)
+{
+    const FlexEGDescription& desc = *desc_;
+    size_t numFrames = out.size();
+    const float samplePeriod = samplePeriod_;
+
+    // Skip the initial delay, for frame-accurate trigger
+    size_t skipFrames = std::min(numFrames, delayFramesLeft_);
+    if (skipFrames > 0) {
+        delayFramesLeft_ -= skipFrames;
+        fill(absl::MakeSpan(out.data(), skipFrames), 0.0f);
+        out.remove_prefix(skipFrames);
+        numFrames -= skipFrames;
+    }
+
+    // Envelope finished?
+    if (currentStageNumber_ >= desc.points.size()) {
+        fill(out, 0.0f);
```

Comment:
All envelopes end at 0? Or should it be `currentLevel_`?

Reply:
Yes envelopes end at zero. Try in ARIA:
```
<region> sample=*sine
//eg15_amplitude=100
eg15_ampeg=1
eg15_time1=.1  eg15_level1=.25
eg15_time2=.15  eg15_level2=1
eg15_time3=.2  eg15_level3=.5 eg15_sustain=3
eg15_time4=.4 eg15_level4=1
```

The envelope will return to 0 after the final ramp.
This will create a very noticeable discontinuity.

![Capture du 2020-09-13 07-59-55](https://user-images.githubusercontent.com/17614485/93011519-2225f080-f597-11ea-9b96-9b8e46fd9531.png)

Reply:
Hmm ok, that actually feels a bit strange